File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline

n = int(input())
board = [[False] * n for _ in range(n)]
j_check = [False] * n
ans = 0

def n_queen(k):
    global ans
    if k == n:
        ans += 1
        return
    for j in range(n):
        if not j_check[j] and check(k,j):
            j_check[j] = True
            board[k][j] = True
            n_queen(k+1)
            j_check[j] = False
            board[k][j] = False
        
def check(x,y):
    i = x
    j = y
    while j > 0 and i > 0:
        i -= 1
        j -= 1
        if board[i][j]:
            return False
    i = x
    j = y
    while j < n - 1 and i > 0:
        i -= 1
        j += 1
        if board[i][j]:
            return False

    return True

n_queen(0)
print(ans)

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static int ans;
	static boolean[] check_col;
	static boolean[] check1;
	static boolean[] check2;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		check_col = new boolean[N];
		check1 = new boolean[2 * N - 1];
		check2 = new boolean[2 * N - 1];
		ans = 0;
		n_queen(0);
		System.out.println(ans);
	}

	static void n_queen(int row) {
		if (row == N) {
			ans++;
			return;
		}
		for (int col = 0; col < N; col++) {
			if (!check_col[col] && !check1[row + col] && !check2[col - row + (N - 1)]) {
				check_col[col] = true;
				check1[row + col] = true;
				check2[col - row + (N - 1)] = true;
				n_queen(row + 1);
				check_col[col] = false;
				check1[row + col] = false;
				check2[col - row + (N - 1)] = false;
			}
		}
	}
}